<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>A Bioinformatics Pipeline for Biologically Relevant HVG Selection in scRNA-seq Data for Training Predictive DL Models</title>

<!--  =================================== 1 =================================  -->
	<!-- Link to Prism.js Theme (Change this line to test different themes) -->
   
   
	<link href="https://cdnjs.cloudflare.com/ajax/libs/prism-themes/1.5.0/prism-ghcolors.min.css" rel="stylesheet">

<!--  ====================================================================  -->

	
  <style>
    body {
      background-color: #f9f9f9;
      font-family: Arial, sans-serif;
      margin: 0;
      padding: 0;
    }
	  
   

.header {
            font-size: 35px;
            background-color: #004080;
            color: #FFFFFF;
            min-height: 160px; /* shrink height to fit content, not fixed */ /* height: 300px;Set an explicit height */
            padding: 80px 20px;
            text-align: center;
            position: relative; /* Add relative positioning here */
	    margin: 40px auto;
            max-width: 1000px;  /* adjust width to match your body content */
            border-radius: 20px;
        }
   

 
}
	   
	  
    .header a {
      color: #fff;
      margin-right: 15px;
      text-decoration: none;
      font-size: 18px;
    }
    .header a:hover {
      text-decoration: underline;
    }
    main {
      max-width: 900px;
      margin: 40px auto;
      background-color: #fff;
      padding: 40px 40px;
      border-radius: 8px;
      box-shadow: 0 0 10px rgba(0, 0, 0, 0.05);
    }
    h1, h2, h3 {
      color: #004080;
    }
    h1 {
      font-size: 28px;
      margin-bottom: 10px;
      color: #FFFFFF; 
    }
    h2 {
      font-size: 22px;
      margin-top: 40px;
      margin-bottom: 10px;
    }
    h3 {
      font-size: 18px;
      margin-top: 30px;
    }
    ul {
      list-style-type: disc;
      margin-left: 20px;
    }
    .summary {
      font-style: italic;
      margin: 20px 0 30px;
    }
    hr {
      margin: 40px 0;
    }
    .highlight {
      background-color: yellow;
    }


	.home-link {
	  position: absolute;
	  bottom: 20px;
	  left: 20px;
	  color: #fff;
	  text-decoration: none;
	  font-size: 20px;
	}   




	  
/*  =============================================================================================================   */ 
/*  ====================== To highlight the figure or caption when a user clicks on the link:  ================   */ 
	    /*  ====================== Step 1: Add a CSS class for highlighting  ================   */ 
       
        .highlight {
	    animation: highlightAnimation 5s ease-out;
	  }
	
	  @keyframes highlightAnimation {
	    from {
	      background-color: yellow;
	    }
	    to {
	      background-color: transparent;
	    }
	  }
       
/*  =============================================================================================================   */ 


   
	/*  =================== the text color when is selected  ==============   */ 
	    /* To change the color of text selection on a webpage (like the blue or yellow highlight  ), 
		we can use the ::selection CSS pseudo-element. */ 
		::selection {
			background-color: #E0003C; /* Change this to any color you prefer (#E0003C or yellow) */
			color: white; /* Change the text color if needed */
		}
	/*  ===================================================================   */ 




	  
        .code-caption {
	    font-size: 16px; /* Slightly smaller than body text, but still readable */
	    font-style: italic;
	    margin-bottom: 1px; /* Reduce the bottom margin */
	    color: #555;
	    text-align: justify;
	}

	 	    

        /* Code Block Style */
        .code-block {
            margin-bottom: 20px;
            font-size: 16px; /* Increased code block font size */
        }

        /* Prism.js Code Block Font Size */ 
        pre[class*="language-"] {
            font-size: 16px; /* Increase font size for Prism.js code blocks */
        }

	/* Figure Style */
	figure {
	    margin-bottom: 30px; /* Adds margin below the figure to create space */
	}

	/* Adjust ul and li styles */
	ul {
	    margin-top: 0 !important;
	    margin-bottom: 0 !important;
	    padding-left: 20px !important; /* Adjust padding */
	}
	
	li {
	    margin-bottom: 5px !important; /* Adjust space between list items */
	}


	  
	  
/* ============== MOBILE RESPONSIVE FIX (start) ============== */
@media only screen and (max-width: 768px) {
	.header {
		font-size: 20px;           /* ‚úÖ Smaller font for mobile */
		height: auto;
		padding: 40px 15px; /* padding: 20px 10px; */
	}

	.header h1 {
		font-size: 22px;
		line-height: 1.4;
		padding: 10px;
		color: white;
	}


	

	
	/* Title Style */
        .title {
            font-size: 26px;
            font-weight: bold;
            margin: 20px 0;
            text-align: center;
        }

	.page-title {
	    margin: 0;
	    font-size: 32px;
	    font-weight: bold;
	    line-height: 1.5;
	    text-shadow: 1px 1px 2px #000; /* optional for glow effect */
	}

	 

	.explanation {
		font-size: 18px;
		padding: 0 10px;
	}

	body {
		margin: 10px;
	}
}
/* ============== MOBILE RESPONSIVE FIX (end) ============== */	    



/* = Start ============= Add a hover effect for better UX-   the image to slightly "lift" on hover (like a card effect) ============== */	
#figure-2-1:hover {
    transform: scale(1.01);
    transition: transform 0.2s ease-in-out;
}
/* = End  ============== Add a hover effect for better UX-   the image to slightly "lift" on hover (like a card effect) ============== */	

	  
  </style>

	
  <script>
    function highlightFigure(id) {
      const el = document.getElementById(id);
      if (el) {
        el.classList.add("highlight");
        setTimeout(() => el.classList.remove("highlight"), 2000);
      }
    }
  </script>
</head>
<body>
	

	<div class="header">
	  <a href="https://mohammadrezamohajeri.github.io/Main_Portfolio/index.html" class="home-link">üè† Home</a>
	  <h1 class="page-title">A Bioinformatics Pipeline for Biologically Relevant HVG Selection in scRNA-seq Data for Training Predictive DL Models</h1>
	</div>

   

	  

	
	
  <main>
    <div class="summary">
      A flexible and biologically informed pipeline that refines highly variable genes from single-cell RNA-seq data to optimize deep learning model training. 
    </div>

    <h2><strong>1. Introduction</strong> </h2>
    <ul>
      <li><strong>HVGs</strong> are defined as genes that exhibit significantly greater <strong>expression variability</strong> across single cells than expected by random technical noise, and this variability may reflect underlying <strong>regulatory mechanisms</strong> or <strong>disease-relevant transcriptional dynamics</strong> (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1" target="_blank">Hafemeister & Satija, 2019</a>; <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC5727406/" target="_blank">Yip et al., 2017</a>). In high-dimensional single-cell transcriptomic datasets, however, <strong>selecting which of these variable genes are most meaningful as input features for machine learning or deep learning (DL) models</strong> poses a significant challenge. Standard approaches typically rely on identifying <strong>the top 2,000 to 5,000 HVGs</strong> based purely on statistical criteria such as <strong>expression variance</strong>. These HVGs are often used as the primary input features for training deep learning models in downstream applications such as phenotype prediction, as shown by <a href="https://doi.org/10.1093/bib/bbae673" target="_blank">Martorell-Marug√°n et al. (2025)</a>. However, this selection can include genes that are either <strong>noisy</strong>, stably expressed <strong>housekeeping genes</strong>, or biologically irrelevant to the disease or treatment under investigation. For example, <strong>ribosomal and mitochondrial genes</strong>‚Äîalthough highly expressed and sometimes variable‚Äîare often associated with general <strong>cellular maintenance</strong> rather than condition-specific responses, and may obscure meaningful biological signals <a href="https://www.embopress.org/doi/full/10.15252/msb.20188746" target="_blank">[Luecken & Theis, 2019]</a>. Similarly, <strong>housekeeping genes</strong>, which are broadly and stably expressed across tissues, are unlikely to distinguish between pathological states <a href="https://doi.org/10.1016/j.tig.2013.05.010" target="_blank">[Eisenberg & Levanon, 2013]</a>. Even genuinely variable genes may fail to capture condition-specific biological signals if they do not align with meaningful biological processes <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6781572/" target="_blank">[Yip et al., 2017]</a>. Therefore, a <strong>biologically guided assessment of HVG sets</strong>‚Äîgrounded in curated GO terms or pathway-based references‚Äîcan significantly improve both the <strong>interpretability and effectiveness of model features</strong>.</li>
      <li>To address this limitation, we developed a bioinformatics pipeline tailored for single-cell RNA-seq (scRNA-seq) data in 10x Genomics (i.e., genes, barcodes, and matrix files). This pipeline integrates <strong>statistical variability</strong> with <strong>biological relevance</strong> by benchmarking HVG sets against a <strong>curated list of genes</strong> known to be involved in context-specific pathways. In doing so, it helps researchers derive more robust and interpretable gene sets for model training.</li>
      <li>The motivation behind this pipeline stems from the need to better align computational <strong>feature selection</strong> with <strong>biological insight</strong>. Instead of relying solely on enrichment analyses post hoc, we proactively incorporate pathway- and GO-term-based biological knowledge derived from literature. This biologically anchored approach increases the interpretability and accuracy of predictive models.</li>
      <li>While the initial use case involved <strong>kidney cancer patients</strong>  treated with <strong>immune checkpoint blockade (ICB) therapy</strong>, the pipeline is designed to be generalizable. Users can adjust pathway and GO term selections to suit other diseases and treatment modalities.</li>
    </ul>

    <h2><strong>2. Pipeline Overview</strong> </h2>
    <ul>
      <li><strong>Step 1:</strong> Extract HVGs from selected cell types and generating multiple HVG sets (e.g. 100 to 9,000 genes) for downstream evaluation.</li>
      <li><strong>Step 2:</strong> Determine the optimal HVG set by assessing when adding more genes yields minimal gain in overlap with immune-related genes extracted via biomaRt and msigdbr using literature-informed GO terms and pathways.</li>
      <li><strong>Step 3:</strong> Generate a refined gene list by intersecting the selected HVG set with immune-related genes derived from literature-informed GO terms and pathways, using biomaRt and msigdbr.</li>
    </ul>

    <!-- --------------------------------------------------------- -->
    <!-- First Figure: 2.1 -->
		    <figure id="figure-2-1" style="text-align: justify; margin: 10px;">
		        <a href="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Bio_Feature_Selection_DL_input.png" target="_blank">
		            <img src="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Bio_Feature_Selection_DL_input.png" 
		                 alt="Sample Distribution by source_name_ch1" 
		                 style="display: block; background-color: #f0f0f0; padding: 10px; border-radius: 15px; 
		                        box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.2); max-width: 100%; height: auto;">
		        </a>
		        <figcaption id="figure-caption-2-1">
		            <strong>Figure 2.1. Schematic diagram of the Bioinformatics Pipeline for Selecting Biologically Relevant HVGs in scRNA-seq to Train Predictive DL Models</strong> 
				  This infographic illustrates a streamlined, user-friendly pipeline developed to extract <strong>highly variable</strong> and <strong>biologically relevant</strong> genes from single-cell RNA-seq data for <strong>machine learning (ML)</strong> and <strong>deep learning (DL) model training</strong>. The pipeline is designed to support <strong>biological insight-driven feature selection</strong>, even for users with minimal coding or biology expertise. 

<ul> <li><b>Step 1: Extract HVGs:</b>  Users start by uploading standard scRNA-seq files (10X format) and selecting specific cell types of interest‚Äîsuch as only immune checkpoint blockade (ICB)-exposed cells or responders & non-responders‚Äîbased on metadata. This tailored inclusion ensures flexibility across research questions. Multiple HVG sets are generated (e.g., 100 to 9,000 genes) to allow evaluation of optimal size later.</li> <li><b>Step 2: Evaluate Biological Relevance of HVG Lists:</b>  Using <strong>biomaRt </strong>and <strong>msigdbr</strong> packages in R, users compile a custom list of biologically relevant genes based on selected <strong>GO terms</strong> and <strong>immune-related pathways</strong>. This list serves as a reference to <strong></strong>strong>biologically assess the generated HVG sets. The pipeline visualizes how overlap changes with increasing HVG count and helps pinpoint an <strong>optimal HVG threshold</strong> that maximizes biological relevance while minimizing redundancy. </li> <li><b>Step 3: Final Gene List Extraction:</b>  From the <strong>optimal HVG set</strong> (e.g., 5,000 genes), the pipeline intersects with the <strong>reference gene list</strong> to extract a final output: genes that are both <strong>highly variable</strong> and <strong>biologically meaningful</strong> - ideal for <strong>predictive modeling</strong>.   </li> <li><b>Key Innovations</b> ‚úî  Integrates <strong>statistical variability</strong> with <strong>biological meaning</strong> ‚úî Enables metadata-based filtering (e.g., ICB-Exposed only cells) ‚úî Supports <strong>user-defined GO terms and pathways</strong> ‚úî Identifies an <strong>optimal HVG cutoff</strong> ‚úî Generates gene lists for DL/ML <strong>model training</strong> ‚úî Designed for users with <strong>minimal coding</strong> or <strong>biology background</strong> & expertise</li> </ul>


				
		        </figcaption>
		    </figure>
     <!-- -------------------------------------------------------- -->
    
   

    
    <h2><strong>3. Step-by-Step Breakdown</strong> </h2>

    <h3> <strong>3.1 Step One: Generate HVG Sets</strong> </h3>
    <ul>
      <li>The pipeline starts with scRNA-seq input files in 10X format. Users can specify which cell populations to include‚Äîfor instance, based on therapy response categories such as partial response (PR), stable disease (SD), or progressive disease (PD). The selected cells are saved into a new directory.</li>
      <li>Using Seurat, the pipeline calculates HVGs and saves them incrementally in sets of 100, from 100 up to 9,000. These sets represent candidate feature spaces for downstream evaluation.</li>
     
    </ul>
	  
 <!-- --------------------------------------------------------- -->
    <!-- First Figure: 3.1 -->
		    <figure id="figure-3-1" style="text-align: justify; margin: 10px;">
		        <a href="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Step_1.png" target="_blank">
		            <img src="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Step_1.png" 
		                 alt="Sample Distribution by source_name_ch1" 
		                 style="display: block; background-color: #f0f0f0; padding: 10px; border-radius: 15px; 
		                        box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.2); max-width: 100%; height: auto;">
		        </a>
		        <figcaption id="figure-caption-3-1">
				
				<ul>
					  <li><strong>Figure 3.1. Step One: Generate HVG Sets:</strong> In this step, we have preprocessing and HVG Identification Workflow Using Seurat package.</li>
					
					  <li><strong>A stepwise outline of the initial preprocessing phase</strong> for single-cell RNA sequencing (scRNA-seq) data using the Seurat package in R. This step forms the foundation of the biologically guided HVG selection pipeline.</li>
					
					  <li><strong>Starting from the standard 10X format</strong> (matrix, barcodes, genes), the workflow involves integrating metadata to annotate cells by phenotypes of interest (e.g., treated, untreated, responders, non-responders, malignant, non-malignant). These annotations guide the selection of relevant cell types to be included in the analysis.</li>
					
					  <li><strong>Following cell selection, quality control (QC) filtering is applied</strong> during the creation of the Seurat object, including steps such as filtering out low-quality cells based on mitochondrial gene content or gene/cell count thresholds.</li>
					
					  <li><strong>The next step includes data normalization and the identification of highly variable genes (HVGs).</strong> Multiple HVG sets are generated by varying the number of top variable genes selected (e.g., 100, 200, ‚Ä¶ up to 9,000), which are saved for downstream biological relevance evaluation and optimal threshold selection.</li>
					
				</ul>





				
		        </figcaption>
		    </figure>
     <!-- -------------------------------------------------------- -->

 

	<!-- -------- START ------------------- Code Block ----------------------------- -->
<div class="code-block">
	    <pre><code class="language-r">
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
	    </code></pre>
	</div>
	<!-- --------- END ------------------ Code Block ----------------------------- -->

      
   
 
    <h3><strong>3.2 Step Two: Assess Biological Relevance of HVG Sets</strong></h3>
    <ul>
      <li>In this step, each HVG set is benchmarked against a reference list of biologically meaningful genes. These genes are extracted using the biomaRt and msigdbr R packages, based on a curated set of GO terms and immunotherapy-relevant pathways identified through literature review.</li>
      <li>The pipeline calculates how many genes in each HVG set overlap with this reference list. This allows us to measure not only variability but also biological interpretability. The goal is to find the "elbow point" or inflection threshold where adding more HVGs no longer substantially increases the number of biologically relevant genes captured.</li>
    </ul>
	  
<!-- --------------------------------------------------------- -->
    <!-- First Figure: 3.2 -->
		    <figure id="figure-3-2" style="text-align: justify; margin: 10px;">
		        <a href="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Step2.png" target="_blank">
		            <img src="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Step2.png" 
		                 alt="Sample Distribution by source_name_ch1" 
		                 style="display: block; background-color: #f0f0f0; padding: 10px; border-radius: 15px; 
		                        box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.2); max-width: 100%; height: auto;">
		        </a>
		        <figcaption id="figure-caption-3-2">

				<ul>
				  	<li><strong>Figure 3.2.</strong> This diagram outlines the process of assessing the biological relevance of each HVG set by benchmarking against curated gene lists (Reference list of genes) (Step 2). The workflow starts with compiling reference gene sets using the <strong>biomaRt</strong> and <strong>msigdbr</strong> packages in R, based on Gene Ontology terms and immunotherapy-related pathways derived from literature. Each HVG set is then evaluated for overlap with this reference list, and the number of overlapping genes is quantified and plotted. This visualization helps identify an optimal threshold‚Äîtypically an inflection point‚Äîwhere increasing the number of HVGs no longer yields substantial gains in biological interpretability.</li>
				</ul>

				
		        </figcaption>
		    </figure>
     <!-- -------------------------------------------------------- -->



	  <!-- -------- START ------------------- Code Block ----------------------------- -->
<div class="code-block">
	    <pre><code class="language-r">
	# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
	    </code></pre>
	</div>
	<!-- --------- END ------------------ Code Block ----------------------------- -->

	  
    <h3>3.3 Step Three: Extract Biologically Relevant HVGs for Model Training</h3>
    <ul>
      <li>After identifying the optimal HVG set (e.g., 5,000 genes), the pipeline intersects it once more with the reference gene list. The result is a refined gene set that is both statistically variable and biologically informed, and is now ready to be used as input for deep learning or machine learning models.</li>
    </ul>
	  <!-- --------------------------------------------------------- -->
    <!-- First Figure: 3.3 -->
		    <figure id="figure-3-3" style="text-align: justify; margin: 10px;">
		        <a href="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Step3.png" target="_blank">
		            <img src="Pictures/HVG_Refiner_Bioinformatics_Pipeline/Step3.png" 
		                 alt="Sample Distribution by source_name_ch1" 
		                 style="display: block; background-color: #f0f0f0; padding: 10px; border-radius: 15px; 
		                        box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.2); max-width: 100%; height: auto;">
		        </a>
		        <figcaption id="figure-caption-3-3">
		            
				<ul>
				 	 <li><strong>Figure 3.3.</strong> <strong>Step Three:</strong> This diagram illustrates the final step of the biologically guided HVG selection pipeline. A previously selected HVG set‚Äîidentified as optimal based on overlap analysis‚Äîis intersected with the curated list of immune-related or biologically meaningful genes. This step refines the HVG set, narrowing it down to genes that are both highly variable and functionally relevant. The resulting gene list serves as the final input for downstream machine learning or deep learning model training.</li>
				</ul>

		        </figcaption>
		    </figure>
     <!-- -------------------------------------------------------- -->


	  <!-- -------- START ------------------- Code Block ----------------------------- -->
<div class="code-block">
	    <pre><code class="language-r">
	# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
# Code will be added here soon
	    </code></pre>
	</div>
	<!-- --------- END ------------------ Code Block ----------------------------- -->

	  
    <h2><strong>4. Final Output and Applicability</strong></h2>
    <ul>
      <li>The final output of the pipeline is a biologically refined list of highly variable genes (HVGs), saved as a .csv file. This gene set serves as a high-quality input for downstream deep learning models aimed at predicting treatment response or other clinically or biologically relevant outcomes ‚Äî depending on the specific context of the dataset.</li>
      <li>One of the core strengths of the pipeline is its ability to determine an optimal HVG threshold. This threshold reflects the point at which adding more genes no longer significantly increases the number of biologically meaningful genes captured. Selecting fewer HVGs than this optimal point may risk excluding informative genes with lower variability that nonetheless play crucial roles in disease mechanisms or treatment responses. Conversely, selecting too many HVGs may introduce noise by including irrelevant genes, which can negatively impact model training and performance.</li>
      <li>To address this, the pipeline doesn't rely solely on variance rankings. Instead, it incorporates biological filtering using pathway- and GO-term-based reference gene sets relevant to the disease or treatment under investigation. This ensures that even HVGs with lower variability ‚Äî but biological significance ‚Äî are retained, while biologically irrelevant genes are excluded. This dual approach improves the interpretability and reliability of feature selection for downstream modeling.
</li>
    </ul>
  </main>

	<!-- Prism.js library for syntax highlighting -->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.28.0/prism.min.js"></script>
	  
    <!-- Load R language support for Prism -->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.28.0/components/prism-r.min.js"></script>


	  
<!--  =============================================================================================================   -->
<!--  ====================== To highlight the figure or caption when a user clicks on the link:  ================   -->
	    <!--  ====================== Step 2: Add JavaScript to apply the class  ================   --> 
	<script>
	  function highlightFigure(figureId) {
	    var figureCaption = document.getElementById("figure-caption-" + figureId);
	    figureCaption.classList.add('highlight');
	    setTimeout(function() {
	      figureCaption.classList.remove('highlight');
	    }, 2000);  // Highlight lasts for 5 seconds
	  }
	</script>  
<!--  =============================================================================================================   -->

	
</body>
</html>
